stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -572,25 +602,62 @@ swift_conformsToSwiftProtocolImpl(const Metadata * const type,
     return nullptr;
   }
 
-  // Really scan conformance records.
-  for (size_t i = startIndex; i < endIndex; i++) {
-    auto &section = snapshot.Start[i];
-    // Eagerly pull records for nondependent witnesses into our cache.
-    for (const auto &record : section) {
-      auto &descriptor = *record.get();
-
-      // We only care about conformances for this protocol.
-      if (descriptor.getProtocol() != protocol)
-        continue;
-
-      // If there's a matching type, record the positive result.
-      ConformanceCandidate candidate(descriptor);
-      if (candidate.getMatchingType(type)) {
-        const Metadata *matchingType = candidate.getConformingTypeAsMetadata();
-        if (!matchingType)
-          matchingType = type;
+  llvm::SmallDenseSet<const ProtocolConformanceDescriptor *, 1>
+    cachedDescriptors;
+  {
+    ScopedLock guard(C.ConformanceDescriptorCacheLock);
+    _scanAdditionalConformanceDescriptors(C);
+    cachedDescriptors = _findConformanceDescriptorsInCache(C, protocol);
+  }
+  
+  bool foundInCache = false;
+  for (auto cachedDescriptor : cachedDescriptors) {
+    ConformanceCandidate candidate(*cachedDescriptor);
+    if (candidate.getMatchingType(type)) {
+      const Metadata *matchingType = candidate.getConformingTypeAsMetadata();
+      if (!matchingType)
+        matchingType = type;
+
+      C.cacheSuccess(matchingType, protocol, cachedDescriptor);
+      foundInCache = true;
+    }
```

Comment:
Huh. I was about to point out that you're missing a `break` here, but I see that we don't have a `break` below either... I suppose we should fix that as a separate thing.